src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,161 @@
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+
+                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
+                Matcher placeholderMatcher = patterns.get(patterns.size() - 1).matcher(rule.getKafkaTopicTemplate());
+                while (placeholderMatcher.find()) {
+                    placeholders.put(placeholderMatcher.group(), null);
+                }
+
+                for (String placeholderKey : mqttTopicPatternParts) {
+                    if (placeholderKey.matches(MQTT_TOPIC_PLACEHOLDER_REGEX) && placeholders.containsKey(placeholderKey)) {
+                        placeholders.put(placeholderKey, matcher.group(removeBrackets(placeholderKey)));
+                    }
+                }
+
+                for (Map.Entry<String, String> entry : placeholders.entrySet()) {
+                    if (entry.getValue() != null) {
+                        mappedKafkaTopic = mappedKafkaTopic.replace(entry.getKey(), entry.getValue());
+                    } else {
+                        throw new IllegalArgumentException("The placeholder " + entry.getKey() + " was not assigned any value.");
+                    }
+                }
+
+                return mappedKafkaTopic;
+            }
+
+        }
+        return DEFAULT_KAFKA_TOPIC;
+    }
+
+    /**
+     * Helper method for Building the regex expressions for the mapping rules.
+     */
+    private void buildRegex() {
+
+        //convert the mqtt patterns to a valid regex expression.
+        // the mqtt pattern can contain placeholders like {something}, + and #.
+        // if the mqtt topic contains a +, we replace it with @singleLevelWildcardRegex
+        // if the mqtt topic contains a #, we replace it with @multiLevelWildcardRegex
+        // if the mqtt topic contains a placeholder (pattern \{\w+\}), we replace it with @placeholderRegex
+        String[] mqttTopicPatternParts;
+        StringBuilder ruleRegex;
+        for (MappingRule rule : this.rules) {
+            mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+            int lastElementIndex = mqttTopicPatternParts.length - 1;
+            ruleRegex = new StringBuilder();
+            for (String part : mqttTopicPatternParts) {
+                if (part.matches(MQTT_TOPIC_PLACEHOLDER_REGEX)) {
+                    ruleRegex.append(buildNamedRegexExpression(part));
+                } else if (part.equals(MQTT_TOPIC_SINGLE_LEVEL_WILDCARD_CHARACTER)) {
+                    ruleRegex.append(SINGLE_LEVEL_WILDCARD_REGEX);
+                } else if (part.equals(MQTT_TOPIC_MULTI_LEVEL_WILDCARD_CHARACTER)) {
+                    ruleRegex.append(MULTIPLE_LEVEL_WILDCARD_REGEX);
+                } else {
+                    ruleRegex.append(part);
+                }
+                if (!part.equals(mqttTopicPatternParts[lastElementIndex])) {
+                    ruleRegex.append("/");
+                }
```

Comment:
I found a bug ... try a test with this rule:

```java
rules.add(new MappingRule("buildings_{building}_rooms_{room}_device_{device}", "buildings/+/rooms/+/device/+"));

assertThrows(IllegalArgumentException.class, () -> mapper.map("buildings/10/rooms/5/device/3"));
```

it doesn't add `/` after finding a `+` because ... `+` is the last element :-)

Reply:
interesting...

Reply:
fixed.